File: Python/0046_permutations.py
# 2020-08-05 13:05:09

class Solution:

    def __init__(self):
        self.ret = []

    def all_permute_from_start(self, nums: List[int], start: int):
        if start == len(nums):
            self.ret.append([x for x in nums])
        for i in range(start, len(nums)):
            nums[i], nums[start] = nums[start], nums[i]
            self.all_permute_from_start(nums, start + 1)
            nums[i], nums[start] = nums[start], nums[i]


    def permute(self, nums: List[int]) -> List[List[int]]:
        self.all_permute_from_start(nums, 0)
        return self.ret

File: TypeScript/0046-permutations.ts
// 2020-08-05 15:36:58

function allPermuteFromStart(nums: number[], start: number, ret: number[][]) {
  if (start == nums.length) {
    ret.push([...nums]);
  } else {
    for (let i = start; i < nums.length; i++) {
      [nums[start], nums[i]] = [nums[i], nums[start]];
      allPermuteFromStart(nums, start + 1, ret);
      [nums[start], nums[i]] = [nums[i], nums[start]];
    }
  }
}

function permute(nums: number[]): number[][] {
  const ret: number[][] = []

  if (nums == null) {
    return [];
  }
  allPermuteFromStart(nums, 0, ret);
  return ret;
}
